shell/platform/linux/fl_event_channel.cc

```diff
@@ -0,0 +1,233 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_event_channel.h"
+
+#include <gmodule.h>
+
+#include "flutter/shell/platform/linux/fl_method_codec_private.h"
+
+static constexpr char kListenMethod[] = "listen";
+static constexpr char kCancelMethod[] = "cancel";
+static constexpr char kEventRequestError[] = "error";
+
+struct _FlEventChannel {
+  GObject parent_instance;
+
+  // Messenger to communicate on.
+  FlBinaryMessenger* messenger;
+
+  // Channel name.
+  gchar* name;
+
+  // Codec to en/decode messages.
+  FlMethodCodec* codec;
+
+  // Function called when the stream is listened to / cancelled.
+  FlEventChannelHandler listen_handler;
+  FlEventChannelHandler cancel_handler;
+  gpointer handler_data;
+  GDestroyNotify handler_data_destroy_notify;
+};
+
+struct _FlEventChannelResponseHandle {
+  GObject parent_instance;
+
+  FlBinaryMessengerResponseHandle* response_handle;
+};
+
+// Added here to stop the compiler from optimising this function away.
+G_MODULE_EXPORT GType fl_event_channel_get_type();
+
+G_DEFINE_TYPE(FlEventChannel, fl_event_channel, G_TYPE_OBJECT)
+
+// Handle method calls from the Dart side of the channel.
+static FlMethodResponse* handle_method_call(FlEventChannel* self,
+                                            const gchar* name) {
+  FlEventChannelHandler handler;
+  if (g_strcmp0(name, kListenMethod) == 0) {
+    handler = self->listen_handler;
+  } else if (g_strcmp0(name, kCancelMethod) == 0) {
+    handler = self->cancel_handler;
+  } else {
+    g_autofree gchar* message =
+        g_strdup_printf("Unknown event channel request '%s'", name);
+    return FL_METHOD_RESPONSE(
+        fl_method_error_response_new(kEventRequestError, message, nullptr));
+  }
+
+  // If not handled, just accept requests.
+  if (handler == nullptr) {
+    return FL_METHOD_RESPONSE(fl_method_success_response_new(nullptr));
+  }
+
+  FlMethodResponse* response = handler(self, self->handler_data);
+  if (response == nullptr) {
+    g_autofree gchar* message =
+        g_strdup_printf("Event channel request '%s' not responded to", name);
+    return FL_METHOD_RESPONSE(
+        fl_method_error_response_new(kEventRequestError, message, nullptr));
+  }
+
+  if (FL_IS_METHOD_SUCCESS_RESPONSE(response)) {
+    if (fl_method_success_response_get_result(
+            FL_METHOD_SUCCESS_RESPONSE(response)) != nullptr) {
+      g_warning("Event channel %s responded to '%s' with value", self->name,
+                name);
+    }
+  } else if (FL_IS_METHOD_ERROR_RESPONSE(response)) {
```

Comment:
Why is this empty? If it's intentional, a comment explaining why in the body would be helpful.

Reply:
Added comment.